File: NS_fluids_lib/LSA_Rayleigh_Python/python/dump_amplitude.py
"""Write the interface-mode amplitude record of a calculation as plain text.

The record is the quantity the dispersion relation predicts, namely the amplitude of the k-Fourier component of the interface radius; see `interface_mode.py` for why a field norm is unsuitable for an interfacial instability. Output is three columns: time, mode amplitude, and mean interface radius. The last is included because a drifting mean signals a mass-conservation error, which would invalidate the growth rate.

The format is byte-for-byte the format written by the Fortran tool `bin/dump_amplitude`, so that the records of the two implementations are interchangeable and the plotting scripts need not distinguish them.

Usage, from amrex_implicit_interfaces/NS_fluids_lib/LSA_Rayleigh_Python:

    python3 python/dump_amplitude.py <rundir> <wavelength> <outfile>
"""

import argparse
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import interface_mode  # noqa: E402


def main(argv=None):
    ap = argparse.ArgumentParser(
        description="interface-mode amplitude record as plain text")
    ap.add_argument("rundir",
                    help="run directory holding the nddataPLT* plotfiles")
    ap.add_argument("wavelength", type=float,
                    help="axial wavelength, equal to ns.yblob of the deck")
    ap.add_argument("outfile", help="destination text file")
    ap.add_argument("--field", default="L0101",
                    help="level-set field carrying the interface")
    ap.add_argument("--harmonic", type=int, default=1,
                    help="Fourier harmonic imposed by the deck")
    ap.add_argument("--prefix", default="nddataPLT",
                    help="plotfile directory prefix written by the solver")
    args = ap.parse_args(argv)

    try:
        times, amps, means = interface_mode.amplitude_series(
            args.rundir, args.wavelength, level_set_field=args.field,
            harmonic=args.harmonic, prefix=args.prefix)
    except FileNotFoundError as exc:
        print(f"ERROR reading the run: {exc}", file=sys.stderr)
        return 1

    with open(args.outfile, "w") as fh:
        fh.write("# time  amplitude  mean_radius\n")
        for t, a, r in zip(times, amps, means):
            fh.write(f"{t:24.16E}{a:24.16E}{r:24.16E}\n")

    print(f"wrote {len(times)} rows to {args.outfile}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: NS_fluids_lib/LSA_Rayleigh_Fortran/src/dump_amplitude.f90
!> Write the interface-mode amplitude record of a calculation as plain text.
!>
!> The record is the quantity the dispersion relation predicts, namely the amplitude of the k-Fourier component of the interface radius; see interface_mode_mod for why a field norm is unsuitable for an interfacial instability. Output is three columns: time, mode amplitude, and mean interface radius. The last of these is included because a drifting mean signals a mass-conservation error, which would invalidate the growth rate.
!>
!> Writing plain text rather than plotting directly keeps this implementation free of any Python dependency; the figure scripts in plotting/ read these files, and they may equally be plotted elsewhere.
!>
!> ./bin/dump_amplitude <rundir> <wavelength> <outfile>
program dump_amplitude
   use interface_mode_mod
   implicit none

   character(len=512) :: rundir, lambda_text, outfile
   real(dp), allocatable :: times(:), amps(:), mean_radius(:)
   real(dp) :: wavelength
   integer :: n, ierr, i, unit_out

   if (command_argument_count() < 3) then
      write(*,'(a)') 'usage: dump_amplitude <rundir> <wavelength> <outfile>'
      stop 1
   end if
   call get_command_argument(1, rundir)
   call get_command_argument(2, lambda_text)
   call get_command_argument(3, outfile)
   read(lambda_text, *) wavelength

   ! The level set of the first material, L0101, carries the interface; the first harmonic is the mode imposed by the deck.
   call amplitude_series(trim(rundir), wavelength, 'L0101', 1, &
                         times, amps, mean_radius, n, ierr)
   if (ierr /= 0) then
      write(*,'(a,i0)') 'ERROR reading the run, ierr=', ierr
      if (ierr == 1) write(*,'(a)') &
         '  no plotfiles found; the solver writes nddataPLT* directories'
      stop 1
   end if

   open(newunit=unit_out, file=trim(outfile), status='replace', &
        action='write')
   write(unit_out,'(a)') '# time  amplitude  mean_radius'
   do i = 1, n
      write(unit_out,'(3es24.16)') times(i), amps(i), mean_radius(i)
   end do
   close(unit_out)

   write(*,'(a,i0,a,a)') 'wrote ', n, ' rows to ', trim(outfile)
end program dump_amplitude
